ext/fg/js/popup-proxy.js

```diff
@@ -51,92 +55,110 @@ class PopupProxy {
-        if (this._id === null) {
-            return false;
+        if (this._getFrameOffset !== null) {
+            await this._updateFrameOffset();
+            [x, y] = this._applyFrameOffset(x, y);
         }
         return await this._invokeHostApi('containsPoint', {id: this._id, x, y});
     }
 
     async showContent(elementRect, writingMode, type=null, details=null) {
-        const id = await this._getPopupId();
-        elementRect = PopupProxy._convertDOMRectToJson(elementRect);
-        return await this._invokeHostApi('showContent', {id, elementRect, writingMode, type, details});
+        let {x, y, width, height} = elementRect;
+        if (this._getFrameOffset !== null) {
+            await this._updateFrameOffset();
+            [x, y] = this._applyFrameOffset(x, y);
+        }
+        elementRect = {x, y, width, height};
+        return await this._invokeHostApi('showContent', {id: this._id, elementRect, writingMode, type, details});
     }
 
     async setCustomCss(css) {
-        const id = await this._getPopupId();
-        return await this._invokeHostApi('setCustomCss', {id, css});
+        return await this._invokeHostApi('setCustomCss', {id: this._id, css});
     }
 
     clearAutoPlayTimer() {
-        if (this._id === null) {
-            return;
-        }
         this._invokeHostApi('clearAutoPlayTimer', {id: this._id});
     }
 
     async setContentScale(scale) {
-        const id = await this._getPopupId();
-        this._invokeHostApi('setContentScale', {id, scale});
+        this._invokeHostApi('setContentScale', {id: this._id, scale});
     }
 
     // Private
 
-    _getPopupId() {
-        if (this._idPromise === null) {
-            this._idPromise = this._getPopupIdAsync();
-        }
-        return this._idPromise;
-    }
-
-    async _getPopupIdAsync() {
-        const {id} = await this._invokeHostApi('getOrCreatePopup', {id: this._id, parentId: this._parentId});
-        this._id = id;
-        return id;
-    }
-
     _invokeHostApi(action, params={}) {
         if (typeof this._parentFrameId !== 'number') {
             return Promise.reject(new Error('Invalid frame'));
         }
         return this._apiSender.invoke(action, params, `popup-proxy-host#${this._parentFrameId}`);
     }
 
-    static _convertDOMRectToJson(domRect) {
-        return {
-            x: domRect.x,
-            y: domRect.y,
-            width: domRect.width,
-            height: domRect.height
+    async _updateFrameOffset() {
+        const firstRun = this._frameOffsetUpdatedAt === null;
+        const expired = firstRun || this._frameOffsetUpdatedAt < Date.now() - PopupProxy._frameOffsetExpireTimeout;
+        if (this._frameOffsetPromise === null && !expired) { return; }
+
+        if (this._frameOffsetPromise !== null) {
+            if (firstRun) {
+                await this._frameOffsetPromise;
+            }
+            return;
+        }
+
+        this._frameOffsetPromise = this._getFrameOffset();
```

Comment:
This could probably be simplified, from line 134-156:

```js
async _updateFrameOffset() {
  // ...

  const promise = this._updateFrameOffsetInner();
  if (firstRun) {
    await promise;
  }
}

async _updateFrameOffsetInner() {
  this._frameOffsetPromise = this._getFrameOffset();
  try {
    const offset = await promise;
    this._frameOffset = offset !== null ? offset : [0, 0];
    this._frameOffsetUpdatedAt = Date.now();
  } catch (e) {
    logError(e);
  } finally {
    this._frameOffsetPromise = null;
  }
}
```

Reply:
You're right, that's a lot cleaner. `Promise`s all the way

Reply:
https://github.com/FooSoft/yomichan/pull/417/commits/883759b9b7f1d3b5ddb8d3f17a29841c83f725dd